### ISSUE TITLE
Infinite loop hang up in YY_INPUT macro if error occurs in reading a file stream

### ISSUE THREAD

**Issue by HattonLe** [westes/flex#681]
YY_INPUT macro never exits the while loop if reading a file causes an error!

The fault occurs within the YY_INPUT macro so I can't step into it to debug easily, but looking thru the source files in this repo the code at fault is in cpp-flex.skl,  line 1602 - "while ( (result = (int) fread(buf, 1, (yy_size_t) max_size, yyin)) == 0 && ferror(yyin)) "

This code will just sit there forever saying I've got an error, thereby hanging your application.

**Comment by Mightyjo** [comment]
The code in that while loop will break out for any error other that EINTR. That includes running out of bytes to read from the file. In this case, EINTR means fread's underlying system call was interrupted by the kernel, but we can try it again. 

Under normal conditions, you'll go through the loop about file_len / max_size times. If you've opened a character device or something similar you probably won't get an EOF so your fread will be block forever waiting for bytes.

Can you share the code for your scanner?

**Comment by westes** [comment]
When reporting that your scanner encounters an error, please provide the scanner. Better yet, a simplified version that only contains the failing code -- a minimal example. Otherwise we cannot reproduce the problem. Nor can we tell if the problem has been fixed by any changes that may be introduced.

**Comment by HattonLe** [comment]
I expanded the macro code so I could gdb into it... actually the failure it can't handle is if yyin is NULL - fread() just hangs.
(I forgot to put my test script file into the folder where the code was expecting it to be, so the file descriptor never opened.
So it's my fault in the first place, but either yyparse() or yyset_in() should have generated an error to indicate yyin is NULL  in order to prevent this hang up from ever occurring.

**Comment by HattonLe** [comment]
Would be nice if the parser warned you of trying to use a NULL yyin.

**Comment by Mightyjo** [comment]
You're right, that's annoying. We may have lost a guard check. I'll take a look.
